### PR TITLE
MAYA-125649 - Build MayaUsd on RHEL8 with new ABI when Maya switches

### PR DESCRIPTION
**Pull request by seando-adsk**
#### MAYA-125320 - Enable UB2 on MacOSX
- Support for RHEL8 with new ABI. If Maya was built with new ABI then MayaUsd must also be.
- Detect if Maya was built with UB2 and issue warning if MayaUsd is UB2, but Maya is not.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -85,6 +85,10 @@ endif()
 include(cmake/utils.cmake)
 find_package(Maya 2018 REQUIRED)
 
+if(APPLE AND BUILD_UB2 AND NOT MAYA_MACOSX_BUILT_WITH_UB2)
+    message(WARNING  "Maya was NOT built with Universal Binary 2")
+endif()
```
Comment (seando-adsk) [inline CMakeLists.txt:90]: Simply issue a warning plugin is being built with UB2 but Maya was not. Should not cause any problems, but its extra compilation not needed.

File: cmake/compiler_config.cmake [hunk 2]
```diff
@@ -106,7 +106,7 @@ function(mayaUsd_compile_config TARGET)
         if(IS_LINUX)
             target_compile_definitions(${TARGET}
                 PRIVATE
-                    _GLIBCXX_USE_CXX11_ABI=0 # USD is built with old ABI
+                    _GLIBCXX_USE_CXX11_ABI=$<IF:$<BOOL:${MAYA_LINUX_BUILT_WITH_CXX11_ABI}>,1,0>
```
Comment (seando-adsk) [inline cmake/compiler_config.cmake:109]: Set the cxx11 abi to same value as the Maya we are compiling with. The two must match.

File: cmake/modules/FindMaya.cmake [hunk 3]
```diff
@@ -12,17 +12,25 @@
...
 # MAYA_ENABLE_NEW_PRIM_DELETE Enable new delete behaviour for delete command
 # MAYA_HAS_DISPLAY_STYLE_ALL_VIEWPORTS Presence of MFrameContext::getDisplayStyleOfAllViewports.
 # MAYA_ARRAY_ITERATOR_DIFFERENCE_TYPE_SUPPORT Presence of maya array iterator difference_type trait
 # MAYA_HAS_GET_MEMBER_PATHS Presence of MFnSet::getMemberPaths
+# MAYA_HAS_DISPLAY_LAYER_API Presence of MFnDisplayLayer
+# MAYA_HAS_NEW_DISPLAY_LAYER_MESSAGING_API Presence of MDisplayLayerMemberChangedFunction
+# MAYA_HAS_RENDER_ITEM_HIDE_ON_PLAYBACK_API Presence of MRenderItem has HideOnPlayback API
 # MAYA_CAMERA_GIZMO_SUPPORT Support for drawing Ufe cameras and lights in the viewport.
+# MAYA_LINUX_BUILT_WITH_CXX11_ABI Maya Linux was built with new cxx11 ABI.
+# MAYA_MACOSX_BUILT_WITH_UB2 Maya OSX was built with Universal Binary 2.
```
Comment (seando-adsk) [inline cmake/modules/FindMaya.cmake:33]: I added these two. Others were existing but missing the doc line here.

File: cmake/modules/FindMaya.cmake [hunk 4]
```diff
@@ -474,6 +482,50 @@ if (MAYA_API_VERSION VERSION_GREATER_EQUAL 20230200)
...
+if(IS_LINUX AND MAYA_Foundation_LIBRARY)
+    # Determine if Maya (on Linux) was built using the new CXX11 ABI.
+    # If yes, then MayaUsd MUST also be built with new ABI.
+    execute_process(
+        COMMAND
+            nm "${MAYA_Foundation_LIBRARY}"
+        COMMAND
+            grep findVariableReplacement
+        COMMAND
+            grep " T "
```
Comment (seando-adsk) [inline cmake/modules/FindMaya.cmake:495]: Something new I learned is that running multiple commands this way sends the output of one to the input of the next. Here running nm grepping for certain text. Then below on the string find look for the cxx11 version. If present it means Maya was compiled with the new ABI.

File: cmake/modules/FindMaya.cmake [hunk 5]
```diff
@@ -474,6 +482,50 @@ if (MAYA_API_VERSION VERSION_GREATER_EQUAL 20230200)
...
+    # Determine if Maya (on OSX) was built with Universal Binary 2 (x86_64 & arm64).
+    # If yes, then MayaUsd can be built with either: Intel, Arm or both.
+    execute_process(
+        COMMAND
+            lipo -archs "${MAYA_Foundation_LIBRARY}"
+        WORKING_DIRECTORY
+            ${MAYA_LIBRARY_DIR}
+        OUTPUT_VARIABLE
+            maya_lipo_output)
+    string(REGEX MATCHALL "(x86_64|arm64)" maya_ub2_match ${maya_lipo_output})
```
Comment (seando-adsk) [inline cmake/modules/FindMaya.cmake:518]: Use regular expression to see and check list output below to see if both these are present. If they are it means Maya was compiled with UB2 enabled.

**Review by boudrey — APPROVED** [review]

**Review by boudrey — APPROVED** [review]

**Pull request merged**